Pluggable project model

matklad opened the pull request:
This is a WIP at abstracting the project model, as proposed in https://github.com/racer-rust/racer/pull/855#issuecomment-392448281.

I hope that it'll solve https://github.com/rust-lang-nursery/rls/issues/956, but I haven't implemented the RLS side of things yet, so there's a chance that this whole approach breaks down.

Nevertheless, I figured I'd just toss it over the fence for now to see what the folks think :)

kngwyu commented:
Great!
I feel we need to implement this kind of feature too, but now I don't have enough time, so really thank you for this PR.

But,  I think it's better to merge this PR after implementing `RacerProjectModelProvider` to RLS.

kngwyu commented on src/racer/project_model.rs:
```diff
@@ -0,0 +1,216 @@
+use std::path::{Path, PathBuf};
+
+pub trait RacerProjectModelProvider {
```
Do we really need `Racer` prefix here?
From other crates it looks `racer::RacerProjectModelProvider`

kngwyu commented on src/racer/project_model.rs:
```diff
@@ -0,0 +1,216 @@
+            if self.get_deps(manifest).is_none() {
+                // cache doesn't exist
+                // calculating depedencies can be bottleneck we use info! here(kngwyu)
+                info!("[get_outer_crates] cache doesn't exist");
```
Could you please change function name here?

kngwyu commented on src/racer/project_model.rs:
```diff
@@ -0,0 +1,216 @@
+                self.cache_deps(manifest, deps_map);
+            }
+            let deps = self.get_deps(manifest).unwrap();
+            debug!("[get_outer_crates] cache exists for manifest",);
```
Same as above, please change function name in debug!.

kngwyu commented on src/racer/project_model.rs:
```diff
@@ -0,0 +1,216 @@
+                let deps_map = self.resolve_dependencies(&manifest)?;
+                self.cache_deps(manifest, deps_map);
+            }
+            let deps = self.get_deps(manifest).unwrap();
```
Why call `self.get_deps` twice here

kngwyu commented:
I left some comments, but for the whole code, it looks good to me.

matklad commented:
The RLS sideof things is https://github.com/rust-lang-nursery/rls/issues/956. It's super ugly atm, but at least it works! Will look into clean-up of the racer implementation next! Thanks for the review.

matklad commented:
Rebased and cleaned-up. [The RLS](https://github.com/rust-lang-nursery/rls/pull/965) side is super incomplete and ugly, but it works, and indeed get's rid of one Cargo :-)

kngwyu commented:
Now looks good enough to merge :+1: 
But I'm going to wait a little, in case you come up with an alternative idea while working on RLS side :slightly_smiling_face:
